### DISCUSSION TITLE
YOLOV8 Fine-Tuning on Custom Dataset

### DISCUSSION THREAD

**Discussion by mroanjos** [ultralytics/ultralytics#19785] [Q&A]
I am new to all this and am currently training a pretrained yolov8m model from ultralytics on a custom dataset.
But I am getting high loss values, specially when it comes to the box loss. Trying to detect very small punctures.
The dataset seems to be ok and the annotations contain the coordinates of the punctures as well as its area.
I think maybe the boxes might be too small.
I am currently using batch=16 and img_size=640 for 100 epochs.

![image](https://github.com/user-attachments/assets/3933b6a8-6975-4d4a-9a27-0af35e29d3b2)

Currently the training is taking a long of time (it wasn't like this before adding augmentation)(>24h).
Training w/ GPU (RTX 4050)

How should I proceed?

![image](https://github.com/user-attachments/assets/aba19943-bff7-4b60-a8e0-407972504b26)
@UltralyticsAssistant @glenn-jocher

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mroanjos, thank you for sharing your training experience with Ultralytics 🚀! It sounds like you're diving deep into fine-tuning YOLO models, which is fantastic. Let’s help you get on the right track.

We recommend reviewing our [Docs](https://docs.ultralytics.com/) for tips and guidance on custom dataset training. You can explore sections like [Python](https://docs.ultralytics.com/usage/python/) or [CLI](https://docs.ultralytics.com/usage/cli/) usage, along with our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/), which may help optimize your training process.

### Recommendations to Move Forward:
- If this is a 🐛 potential bug, could you please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/)? This will help us investigate further.
- For ❓ custom training questions, please ensure you include as much detail as possible, such as:
    - Examples of your dataset images and labels.
    - Training logs or outputs.
    - Any unique augmentations or hyperparameter modifications you’ve applied.
- Detecting very small objects like punctures can be challenging. You can try experimenting with higher-resolution images or smaller anchor sizes and refer to the [model training tips](https://docs.ultralytics.com/guides/model-training-tips/) to refine your process.

### Accelerating Training 🏎️
Training time can be affected by factors like augmentation complexity or batch size. Ensure your environment is optimized:
- Use the latest version of `ultralytics` by upgrading with:
    ```bash
    pip install -U ultralytics
    ```
- Verify that your GPU is fully utilized and that dependencies are up-to-date, including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/).

### Environments
To simplify training and testing, you can also utilize verified environments:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Docker Images**: See our [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>.

### Community
Join the Ultralytics community to discuss your project and gain insights from other users:
- [Discord](https://discord.com/invite/ultralytics) 🎧
- [Discourse](https://community.ultralytics.com/)
- [Subreddit](https://reddit.com/r/Ultralytics)

An Ultralytics engineer will review your issue and provide additional assistance shortly. Thank you for your patience and for being part of the Ultralytics community! 😊

**Comment by glenn-jocher** [comment]
@mroanjos thank you for sharing your training details. For small object detection like punctures, we recommend:  
1. Increasing image size to 1280 (add `imgsz=1280`) to improve small object visibility in feature maps  
2. Reducing batch size to 8 or 4 to accommodate larger images on your GPU  
3. Disabling mosaic augmentation (`augment=False`) which can degrade small object performance  

You might also benefit from our YOLO Common Issues guide which covers anchor tuning and small object optimizations. Let us know if these adjustments help improve your box losses.

**Reply by mroanjos** [reply]
Thank you a lot for the quick reply and for the help! @glenn-jocher

**Reply by glenn-jocher** [reply]
You're welcome! We're glad the suggestions helped. Feel free to reach out if you have any other questions as you continue working with your YOLO model - the whole YOLO community here is always happy to help troubleshoot. 🚀

**Reply by mroanjos** [reply]
Hey again, I've made some progress and wanted to check if the metrics seem to be ok or just to see if I am on the right track.
For additional context, my dataset is partitioned 70/20/10.

I've finally got loss values <1.

After my last train this is what I got:
Precision -> 88.6%; Recall -> 82.3%; mAP50 -> 88.1%; mAP50-95 -> 63,9% 

train: box_loss-> 0.8299 ; cls_loss -> 0.5197 ; dfl_loss -> 0.8774
val: box_loss-> 0.988 ; cls_loss -> 0.634 ; dfl_loss -> 0.927

@UltralyticsAssistant @glenn-jocher

**Reply by glenn-jocher** [reply]
@mroanjos your metrics look promising! With mAP50 at 88.1% and stable losses, you're on track. To further improve mAP50-95 (63.9%), consider analyzing misclassified cases in your validation set using the [YOLO Performance Metrics guide](https://docs.ultralytics.com/guides/yolo-performance-metrics) to refine localization accuracy. Keep us posted! 🚀

**Reply by mroanjos** [reply]
Hello again, @glenn-jocher @UltralyticsAssistant 

I've worked a bit more on the fine-tuning and now I've got:
Precision -> 87.4%; Recall -> 84.6%; mAP50 -> 89.9%; mAP50-95 -> 72.4%
and a fitness score of 0.74

train: box_loss-> 0.1898 ; cls_loss -> 0.1359 ; dfl_loss -> 0.7583
val: box_loss-> 0.8801 ; cls_loss -> 0.5903 ; dfl_loss -> 0.9604

I got the mAP50-95 to improve a bit.

**Reply by glenn-jocher** [reply]
@mroanjos That's excellent progress! Your mAP50-95 has improved significantly from 63.9% to 72.4%, and the fitness score of 0.74 is quite good for a custom small object detection task. The dramatic reduction in training box_loss (from 0.8299 to 0.1898) shows your model is learning effectively.

I notice there's still a gap between training and validation losses, which might indicate some overfitting. If you want to push performance further, you could try:

1. Implementing a learning rate scheduler with warmup
2. Experimenting with image tiling for better small object detection as mentioned in our [troubleshooting guide](https://docs.ultralytics.com/guides/yolo-common-issues/)
3. Adding more diverse samples to your training set if possible

These results are already quite strong for detecting small punctures - you're definitely on the right track!

**Reply by mroanjos** [reply]
@UltralyticsAssistant @glenn-jocher 

I am trying to check the model's performance on a test set but I keep getting 0 in every metric even though I can see the model working when I run inference and see it detecting the punctures.

![image](https://github.com/user-attachments/assets/6f4ac4d0-2607-48e9-bd49-0e194d3d7906)


Could you maybe help?
I've checked the labels and everything seems to be in order.
What could be causing this? Overfitting?

**Reply by glenn-jocher** [reply]
@mroanjos This is a common issue when running validation on test sets - seeing zeros in metrics despite successful visual inference. A few likely causes:

1. **Label path mismatch**: Make sure your test labels are in the correct folder structure (typically images in 'test/images/' and labels in 'test/labels/')

2. **Class ID mismatch**: Verify the class IDs in your test labels match those used during training

3. **Confidence threshold**: Try lowering the confidence threshold for validation with `conf=0.1` 

4. **Data.yaml configuration**: Ensure your data.yaml file correctly points to the test set location

You can diagnose this by examining one specific test example - run inference on it, then manually check if its corresponding label file exists and has the correct format. If the model is correctly detecting punctures visually but metrics show zero, it's almost certainly a path or format issue rather than an overfitting problem.
